controllers/front/OrderConfirmationController.php

```diff
@@ -58,17 +61,21 @@ public function init()
         $this->secure_key = Tools::getValue('key', false);
         $order = new Order((int) ($this->id_order));
 
-        if (!$this->id_order || !$this->id_module || !$this->secure_key || empty($this->secure_key)) {
+        if (!$this->id_order || (!$this->isFreeOrder && !$this->id_module) || !$this->secure_key || empty($this->secure_key)) {
```

Comment:
```suggestion
        if (!$this->id_order || (!$order->currentOrder && !$this->id_module) || !$this->secure_key || empty($this->secure_key)) {
```

Reply:
I'm not sure about it, we need to have different behavior if order is free because there's no `$this->id_module` then

Reply:
The confusion might come because the checks being done are clear
```
if (!$this->id_order || (!$this->isFreeOrder && !$this->id_module) || !$this->secure_key || empty($this->secure_key)) {

```
but the **meaning** of these checks is not obvious.

May I suggest using either phpDoc or variable names to clear the confusion? For example

```suggestion
        if ($thereisNoOrder || $orderIsNotFree || orderIsNotSecure) {
```

This allows to remove the `!` that are harder to read when the statement becomes long, and eases the understanding of the condition (right now it's a 4-parts OR statement with 1 statement being an AND statement).

Reply:
@matks I think I'll move these checks to dedicated methods to have it as readable as possible